README.md
# Time Bandit

![header_image](static/header_image.jpg)

[![netlify badge](https://api.netlify.com/api/v1/badges/416b8ca3-82db-470f-9adf-a6d06264ca75/deploy-status)](https://app.netlify.com/sites/mystifying-keller-ab5658/deploys) [![Build Status](https://travis-ci.com/skeptycal/.dotfiles.svg?branch=dev)](https://travis-ci.com/skeptycal/.dotfiles)

## Timer utilities for Python 3.5+ Development

[![macOS Version](https://img.shields.io/badge/macOS-10.16%20BigSur-blue?logo=apple)](https://www.apple.com) [![GitHub Pipenv locked Python version](https://img.shields.io/badge/Python-3.9-yellow?color=3776AB&logo=python&logoColor=yellow)](https://www.python.org/) [![nuxt.js](https://img.shields.io/badge/nuxt.js-2.14.0-35495e?logo=nuxt.js)](https://nuxtjs.org/)

These are handy automation and informational utilities. Add more, change some, make some additions/corrections...

**Please feel free to offer suggestions and [changes][repo-issues]**.

> Copyright © 1976-2020 [Michael Treanor](https:/skeptycal.github.com)

[![License](https://img.shields.io/badge/License-MIT-darkblue)][skep-mit]

## Installation

### TLDR: clone the repo and run `./setup.sh`

## WHY?

-   comparison of different methods
-   profiling features under real world loads with decorators and log readers
-   visualize outcomes in unique ways

---

[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/3454/badge)](https://bestpractices.coreinfrastructure.org/projects/3454) [![test coverage](https://img.shields.io/badge/test_coverage-100%25-6600CC.svg?logo=Coveralls&color=3F5767)](https://coveralls.io)

---

## Prerequisites

-   Python version 3.5+

### Recommended IDE setup:

-   [VSCode][get-code] IDE
-   [Sarah Drasner][sdras]'s Vue VSCode [Extension Pack][sdras-pack]
-   [Don Jayamanne][djay]'s Python [Extension Pack][djay-pack]

### Installed by this setup as needed:

-   [Homebrew][brew]
-   GNU [coreutils][coreutils] for macOS (brew install coreutils)
-   [Pre-Commit][pre-commit] for automated checks
-   [Poetry][poetry] for dependency management, building, publishing, and versioning

## Install

    pip install timebandit

    - or -

if you wish to make modifications or contribute to the open source project:

```sh
# clone the repo
git clone https://www.github.com/skeptycal/

# change to the repo directory
cd user_bin_dir_repo

# run the init script
./init

# optional: # use './init --nobrew' to skip install of homebrew and utilities
./init --nobrew
```

---

## Usage

```py
from timebandit import timeit

# as a decorator

@timeit
def myfunction(x):
    ... do stuff ...

#     - or -

# as a function call

the_result = timeit(myfunction(5))
print(the_result)
```

---

## Feedback

[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-v1.4%20adopted-ff69b4.svg)](CODE_OF_CONDUCT.md)

Suggestions/improvements are [welcome][repo-issues]!

---

## Author

[![twitter/skeptycal](https://s.gravatar.com/avatar/b939916e40df04f870b03e0b5cff4807?s=80)](http://twitter.com/skeptycal 'Follow @skeptycal on Twitter')

[**Michael Treanor**][me]

![Twitter Follow](https://img.shields.io/twitter/follow/skeptycal.svg?style=social) ![GitHub followers](https://img.shields.io/github/followers/skeptycal.svg?label=GitHub&style=social)

[repo-issues]: (https://github.com/skeptycal/dotfiles/issues)
[repo-fork]: (https://github.com/skeptycal/dotfiles/fork)
[me]: (https://www.skeptycal.com)
[skep-image]: (https://s.gravatar.com/avatar/b939916e40df04f870b03e0b5cff4807?s=80)
[skep-twitter]: (http://twitter.com/skeptycal)
[skep-mit]: (https://skeptycal.mit-license.org/1976/)
[mb]: (https://mathiasbynens.be/)
[sdras]: (https://sarahdrasnerdesign.com/)
[djay]: (https://github.com/DonJayamanne)
[get-code]: (https://code.visualstudio.com/download)
[brew]: (https://brew.sh/)
[djay-pack]: (https://marketplace.visualstudio.com/items?itemName=donjayamanne.python-extension-pack)
[sdras-pack]: (https://marketplace.visualstudio.com/items?itemName=sdras.vue-vscode-extensionpack)
[pre-commit]: (https://pre-commit.com/)
[xcode]: (https://developer.apple.com/xcode/)
[coreutils]: (https://www.gnu.org/software/coreutils/)
[poetry]: (https://python-poetry.org/)


CHANGELOG.md
# TimeBandit Changelog

## UNRELEASED

-   Initial Testing
-   add tests
-   add helper functions

## TimeBandit 0.1.0

-   Add type hints
-   Cleanup initial commit
-   update requirements and pyproject.toml

## TimeBandit 0.1.0

-   Initial upload
-   Modified version of standard library `timeit` module
-   Developed on `Python 3.8.5`
-   Dev Release. Not Stable


tests/test_pytest.py

import timebandit


def _add(a: int, b: int) -> (int):
    return a + b


def test_timebandit_basics():

    assert isinstance(timebandit.timeit(func=list), float)


def test_timebandit_repeat():
    assert isinstance(timebandit.repeat(func=set), list)


def test_add():
    for i in range(100):
        a = 3 * i
        b = 4 * i
        assert a + b == _add(a, b)


def test_true():
    """
    Test Doc: Line 1 ...
    """
    assert True


setup.sh
#!/usr/bin/env zsh
# -*- coding: utf-8 -*-
    # shellcheck shell=bash
    # shellcheck source=/dev/null
    # shellcheck disable=2178,2128,2206,2034

#* ###################### Setup

#? ####################### debug
    # SET_DEBUG: set to 1 for verbose testing;
    SET_DEBUG=1

#? ######################## configuration
	[[ ${SHELL##*/} == 'zsh' ]] && set -o shwordsplit
    export YEAR=$(date +%Y)
    t0=$(date +%s.%n)
    _set_basic_colors() {
        export MAIN='\001\033[38;5;229m'
        export WARN='\001\033[38;5;203m'
        export COOL='\001\033[38;5;38m'
        export BLUE='\001\033[38;5;38m'
        export GO='\001\033[38;5;28m'
        export LIME='\001\033[32;1m'
        export CHERRY='\001\033[38;5;124m'
        export CANARY='\001\033[38;5;226m'
        export ATTN='\001\033[38;5;178m'
        export PURPLE='\001\033[38;5;93m'
        export RAIN='\001\033[38;5;93m'
        export WHITE='\001\033[37m'
        export RESTORE='\001\033[0m\002'
        export RESET='\001\033[0m'
    	}
    # ssm - standard script modules or alternate
    . $(which ssm) || _set_basic_colors
    cleanup() {
        # cleanup and exit script
        unset echo

        # calculate and display script time
        t1=$(date +%s.%n)
        dt=$((t1-t0))
        printf '\n%bScript %s took %.3f seconds to load.\n\n' "${GO:-}" "$SCRIPT_NAME" "$dt"
        unset t0 t1 dt
    	}
    trap cleanup EXIT

	NO_BREW=0

    BASH_SOURCE="${0}"
    SCRIPT_NAME="${BASH_SOURCE##*/}"
    SCRIPT_PATH="${BASH_SOURCE%/*}"

    REPO_PATH="$PWD"
    REPO_NAME="${PWD##*/}"
#? ######################## utilities
    # echo () {
    #     if [ -n "$1" ]; then
    #         printf '%s' "$1"
    #         shift
    #     fi
    #     for arg in "$@"; do
    #         printf ' %s' "$arg"
    #     done
    #     printf '%s\n' ''
    # 	}
	dbecho() { (( SET_DEBUG==1 )) && echo "${ATTN:-}$@"; }
    _debug_show_paths() {
        tmp='BASH_SOURCE SCRIPT_NAME SCRIPT_PATH REPO_PATH REPO_NAME TEMPLATE_PATH'
        for i in $tmp; do
            echo "${MAIN:-}${(r:15:)i} ... ${CANARY:-}${(P)i}${RESET:-}"
        done;
    	}

	die() {
		echo "${ATTN:-}$@"
		exit 1
		}

	is_empty() {
		[ -z "$(ls -A $1)" ]
		}

    exists() { command -v $1 > /dev/null 2>&1 ; }
	version() { cat pyproject.toml | grep 'version = '; }

#? ######################## setup
    _setup_brew() {
		if (( $SET_DEBUG == 1 )); then
			echo "${ATTN:-}Install brew if it is not installed..."
		else
        	exists brew || /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"

			brew cleanup
			brew doctor
			brew update

			brew install git hub python@3.8 poetry pre-commit
		fi

    } # > /dev/null 2>&1

	_setup_repo() {
        python3 -m venv .venv
		source ./.venv/bin/activate

		pip install -U pip

        REQS='loguru wheel'
        DEV_REQS='pip wheel setuptools pylint pytest tox coverage pytest-cov'
        DEV_OPTS='flake8 tox-travis Sphinx sphinx-autobuild sphinx-rtd-theme'

		poetry add $REQS
		poetry add --dev $DEV_REQS
		poetry add --dev --optional $DEV_OPTS

		poetry update
		poetry build
		poetry install

	}

    main() {
		(( SET_DEBUG==1 )) && _debug_show_paths
		TESTPATH=$(realpath $PWD)
		while (( $# > 0 )); do
			case $1 in
				-h|--help)
					echo "Usage: $0 [-h|--help] [-v|--version] [--nobrew] "
					exit 0
					;;
				-v|--version)
					echo "$SCRIPT_NAME $(version)"
					exit 0
					;;
				--nobrew)
					NO_BREW=1
					;;
				*)
					echo ""
					;;
			esac
			shift
		done

		(( NO_BREW==1 )) && ( echo 'Skipping HomeBrew install ...'; ) || _setup_brew
        _setup_repo
    }

main "$@"

# ------------------------ stuff ------------------------
#? ${PATH//:/\\n}    - replace all colons with newlines to display as a list
#? ${PATH// /}       - strip all spaces
#? ${VAR##*/}        - return only final element in path (program name)
#? ${VAR%/*}         - return only path elements in path (without program name)


timebandit/_cli.py
#! /usr/bin/env python3
""" Main program, used when run as a script.

    The optional 'args' argument specifies the command line to be parsed,
    defaulting to sys.argv[1:].

    The return value is an exit code to be passed to sys.exit(); it
    may be None to indicate success.

    When an exception happens during timing, a traceback is printed to
    stderr and the return value is 1.  Exceptions at other times
    (including the template compilation) are not caught.

    '_wrap_timer' is an internal interface used for unit testing.  If it
    is not None, it must be a callable that accepts a timer function
    and returns another timer function (used for unit testing).
    """
from timeit import default_timer, Timer, timeit, repeat, default_repeat
import gc
import sys
import time
import itertools
from loguru import logger

LOGURU_LEVEL = 1

logger.info(f'logger initialized: {logger=}')


def main(args=None, *, _wrap_timer=None):
    """ Main program, used when run as a script.

        The optional 'args' argument specifies the command line to be parsed,
        defaulting to sys.argv[1:].

        The return value is an exit code to be passed to sys.exit(); it
        may be None to indicate success.

        When an exception happens during timing, a traceback is printed to
        stderr and the return value is 1.  Exceptions at other times
        (including the template compilation) are not caught.

        '_wrap_timer' is an internal interface used for unit testing.  If it
        is not None, it must be a callable that accepts a timer function
        and returns another timer function (used for unit testing).
        """
    if args is None:
        args = sys.argv[1:]
    import getopt
    try:
        opts, args = getopt.getopt(args, "n:u:s:r:tcpvh",
                                   ["number=", "setup=", "repeat=",
                                    "time", "clock", "process",
                                    "verbose", "unit=", "help"])
    except getopt.error as err:
        print(err)
        print("use -h/--help for command line help")
        return 2

    timer = default_timer
    stmt = "\n".join(args) or "pass"
    number = 0  # auto-determine
    setup = []
    repeat = default_repeat
    verbose = 0
    time_unit = None
    units = {"nsec": 1e-9, "usec": 1e-6, "msec": 1e-3, "sec": 1.0}
    precision = 3
    for o, a in opts:
        if o in ("-n", "--number"):
            number = int(a)
        if o in ("-s", "--setup"):
            setup.append(a)
        if o in ("-u", "--unit"):
            if a in units:
                time_unit = a
            else:
                print("Unrecognized unit. Please select nsec, usec, msec, or sec.",
                      file=sys.stderr)
                return 2
        if o in ("-r", "--repeat"):
            repeat = int(a)
            if repeat <= 0:
                repeat = 1
        if o in ("-p", "--process"):
            timer = time.process_time
        if o in ("-v", "--verbose"):
            if verbose:
                precision += 1
            verbose += 1
        if o in ("-h", "--help"):
            print(__doc__, end=' ')
            return 0
    setup = "\n".join(setup) or "pass"

    # Include the current directory, so that local imports work (sys.path
    # contains the directory of this script, rather than the current
    # directory)
    import os
    sys.path.insert(0, os.curdir)
    if _wrap_timer is not None:
        timer = _wrap_timer(timer)

    t = Timer(stmt, setup, timer)
    if number == 0:
        # determine number so that 0.2 <= total time < 2.0
        callback = None
        if verbose:
            def callback(number, time_taken):
                msg = "{num} loop{s} -> {secs:.{prec}g} secs"
                plural = (number != 1)
                print(msg.format(num=number, s='s' if plural else '',
                                 secs=time_taken, prec=precision))
        try:
            number, _ = t.autorange(callback)
        except:
            t.print_exc()
            return 1

        if verbose:
            print()

    try:
        raw_timings = t.repeat(repeat, number)
    except:
        t.print_exc()
        return 1

    def format_time(dt):
        unit = time_unit
        scale: float = 1.0

        if unit is not None:
            scale = units[unit]
        else:
            scales = [(scale, unit) for unit, scale in units.items()]
            scales.sort(reverse=True)
            for scale, unit in scales:
                if dt >= scale:
                    break

        return "%.*g %s" % (precision, dt / scale, unit)

    if verbose:
        print("raw times: %s" % ", ".join(map(format_time, raw_timings)))
        print()
    timings = [dt / number for dt in raw_timings]

    best = min(timings)
    print("%d loop%s, best of %d: %s per loop"
          % (number, 's' if number != 1 else '',
             repeat, format_time(best)))

    best = min(timings)
    worst = max(timings)
    if worst >= best * 4:
        import warnings
        warnings.warn_explicit("The test results are likely unreliable. "
                               "The worst time (%s) was more than four times "
                               "slower than the best time (%s)."
                               % (format_time(worst), format_time(best)),
                               UserWarning, '', 0)
    return None


timebandit/__init__.py
#!/usr/bin/env python3 -m

""" ### *** This is not the original python source code ***
    This is a variation on the python3 timeit function from python 3.8.5 and
    is covered by the PSF license. A copy is included at the end of this file
    or it can be found here: https://docs.python.org/3/license.html


    ####* Original (modified) docstring starts here:
    Tool for measuring execution time of small code snippets.

    This module avoids a number of common traps for measuring execution
    times.  See also Tim Peters' introduction to the Algorithms chapter in
    the Python Cookbook, published by O'Reilly.

    Library usage: see the Timer class.

    Command line usage:
        python timeit.py [-n N] [-r N] [-s S] [-p] [-h] [--] [statement]

    Options:
    -n/--number N: how many times to execute 'statement' (default: see below)
    -r/--repeat N: how many times to repeat the timer (default 5)
    -s/--setup S: statement to be executed once initially (default 'pass').
                    Execution time of this setup statement is NOT timed.
    -p/--process: use time.process_time() (default is time.perf_counter())
    -v/--verbose: print raw timing results; repeat for more digits precision
    -u/--unit: set the output time unit (nsec, usec, msec, or sec)
    -h/--help: print this usage message and exit
    --: separate options from statement, use when statement starts with -
    statement: statement to be timed (default 'pass')

    A multi-line statement may be given by specifying each line as a
    separate argument; indented lines are possible by enclosing an
    argument in quotes and using leading spaces.  Multiple -s options are
    treated similarly.

    If -n is not given, a suitable number of loops is calculated by trying
    increasing numbers from the sequence 1, 2, 5, 10, 20, 50, ... until the
    total time is at least 0.2 seconds.

    Note: there is a certain baseline overhead associated with executing a
    pass statement.  It differs between versions.  The code here doesn't try
    to hide it, but you should be aware of it.  The baseline overhead can be
    measured by invoking the program without arguments.

    Classes:

        Timer

    Functions:

        timeit(string, string) -> float
        repeat(string, string) -> list
        default_timer() -> float

    """


import sys
from pathlib import Path

from loguru import logger
from timebandit._cli import main as CLI
from timebandit.timeit import *

# from .timeit import *


__version__: str = None

logger.info(f"filename: {Path(__file__).name}")
logger.info(f"{Path().cwd().parent=}")
logger.info("version: ", __version__)

# for line in Path('../pyproject.toml').open(mode='rt').readlines():
#     if line.startswith("version = "):
#         __version__ = line.split('"')[1]
#         break

if not __version__:
    __version__ = 'unknown'

if __name__ == "__main__":

    sys.exit(CLI())


# ------------> example from pluggy project
# *****************************************************************

# from .hooks import HookspecMarker,  HookimplMarker
# from .callers import HookCallError
# from .manager import PluginManager, PluginValidationError

# try:
#     from ._version import version as __version__
# except ImportError:
#     # broken installation, we don't even try
#     # unknown only works because we do poor mans version compare
#     __version__ = "unknown"

# __all__ = [
#     "PluginManager",
#     "PluginValidationError",
#     "HookCallError",
#     "HookspecMarker",
#     "HookimplMarker",
# ]


timebandit/timeit.py
#! /usr/bin/env python3
""" ### *** This is not the original python source code ***
    This is a variation on the python3 timeit function from python 3.8.5 and
    is covered by the PSF license. A copy is included at the end of this file
    or it can be found here: https://docs.python.org/3/license.html


    ####* Original (modified) docstring starts here:
    Tool for measuring execution time of small code snippets.

    This module avoids a number of common traps for measuring execution
    times.  See also Tim Peters' introduction to the Algorithms chapter in
    the Python Cookbook, published by O'Reilly.

    Library usage: see the Timer class.

    Command line usage:
        python timeit.py [-n N] [-r N] [-s S] [-p] [-h] [--] [statement]

    Options:
    -n/--number N: how many times to execute 'statement' (default: see below)
    -r/--repeat N: how many times to repeat the timer (default 5)
    -s/--setup S: statement to be executed once initially (default 'pass').
                    Execution time of this setup statement is NOT timed.
    -p/--process: use time.process_time() (default is time.perf_counter())
    -v/--verbose: print raw timing results; repeat for more digits precision
    -u/--unit: set the output time unit (nsec, usec, msec, or sec)
    -h/--help: print this usage message and exit
    --: separate options from statement, use when statement starts with -
    statement: statement to be timed (default 'pass')

    A multi-line statement may be given by specifying each line as a
    separate argument; indented lines are possible by enclosing an
    argument in quotes and using leading spaces.  Multiple -s options are
    treated similarly.

    If -n is not given, a suitable number of loops is calculated by trying
    increasing numbers from the sequence 1, 2, 5, 10, 20, 50, ... until the
    total time is at least 0.2 seconds.

    Note: there is a certain baseline overhead associated with executing a
    pass statement.  It differs between versions.  The code here doesn't try
    to hide it, but you should be aware of it.  The baseline overhead can be
    measured by invoking the program without arguments.

    Classes:

        Timer

    Functions:

        timeit(string, string) -> float
        repeat(string, string) -> list
        default_timer() -> float

    """

import gc
import sys
import time
import itertools
import typing as t
from loguru import logger

__all__ = ["Timer", "timeit", "repeat", "default_timer", "default_repeat"]

dummy_src_name: str = "<timeit-src>"
default_number: int = 1000000
default_repeat: int = 5
default_timer: time.perf_counter = time.perf_counter()

logger.info('timeit initialized')

_globals: t.Dict[str, t.Any] = globals

# Don't change the indentation of the template; the reindent() calls
# in Timer.__init__() depend on setup being indented 4 spaces and stmt
# being indented 8 spaces.
template: str = """
def inner(_it, _timer{init}):
    {setup}
    _t0 = _timer()
    for _i in _it:
        {stmt}
    return _timer() - _t0
"""


def reindent(src, indent):
    """Helper to reindent a multi-line statement."""
    return src.replace("\n", "\n" + " "*indent)


def _pass():
    """ Helper to provide a callable form of doing nothing ... 'pass' """
    pass


class Timer:
    """ Modified to accept only callable functions, not code snippets
        passed as strings.

        Class for timing execution speed of functions.

        Parameters:

        - func: the function to be timed
        - setup: a setup function that will be run before timing
            begins (default: '_pass')
        - timer: (optional) the timer used for timing
        - globals: (optional) namespace to be used

        'func' and 'setup' default to 'pass()'; the 'timer' function
        is platform-dependent (see module doc string).  If 'globals'
        is specified, the code will be executed within that namespace
        (as opposed to inside timeit's namespace).

        To measure the execution time of the function, use the
        timeit() method.  The repeat() method is a convenience to call
        timeit() multiple times and return a list of results.

        """

    def __init__(self, func: t.Callable = None,
                 setup: t.Callable = None,
                 timer: time.perf_counter = default_timer,
                 globals: t.Dict[str, t.Any] = None):
        """Constructor.  See class doc string. """
        src: str = ''
        init: str = ''
        local_ns: t.Dict = {}
        global_ns: t.Dict[str, t.Any] = _globals(
        ) if globals is None else globals

        logger.info(f"{src=}")
        logger.info(f"{init=}")
        logger.info(f"{local_ns=}")

        self.setup = setup if setup else _pass
        self.func = func if func else _pass
        self.timer: time.perf_counter = timer

        logger.info(f"{self.func=}")
        logger.info(f"{self.setup=}")
        logger.info(f"{self.timer=}")

        if callable(self.setup):
            local_ns['_setup'] = self.setup
            init += ', _setup=_setup'
            setup = '_setup()'
        else:
            raise ValueError("setup is not callable")
        if callable(self.func):
            local_ns['_func'] = self.func
            init += ', _func=_func'
            stmt = '_func()'
        else:
            raise ValueError("func is not callable")
        src = template.format(stmt=stmt, setup=setup, init=init)
        self.src = src  # Save for traceback display
        code: str = compile(src, dummy_src_name, "exec")
        exec(code, global_ns, local_ns)
        self.inner = local_ns["inner"]

    def print_exc(self, file: t.Union[t.IO[str], None] = None) -> None:
        """Helper to print a traceback from the timed code.

            Typical use:

                t = Timer(...)       # outside the try/except
                try:
                    t.timeit(...)    # or t.repeat(...)
                except:
                    t.print_exc()

            The advantage over the standard traceback is that source lines
            in the compiled template will be displayed.

            The optional file argument directs where the traceback is
            sent; it defaults to sys.stderr.
            """
        import linecache
        import traceback
        if self.src is not None:
            linecache.cache[dummy_src_name] = (len(self.src),
                                               None,
                                               self.src.split("\n"),
                                               dummy_src_name)
        # else the source is already stored somewhere else

        traceback.print_exc(file=file)

    def timeit(self, number: int = default_number) -> float:
        """ Time 'number' executions of the main statement.

            To be precise, this executes the setup statement once, and
            then returns the time it takes to execute the main statement
            a number of times, as a float measured in seconds.  The
            argument is the number of times through the loop, defaulting
            to one million.  The main statement, the setup statement and
            the timer function to be used are passed to the constructor.
            """
        it: t.Iterator[int] = itertools.repeat(None, number)
        gcold: bool = gc.isenabled()
        gc.disable()
        try:
            timing: float = self.inner(it, self.timer)
        finally:
            if gcold:
                gc.enable()
        return timing

    def repeat(self, repeat=default_repeat, number=default_number):
        """ Call timeit() a few times.

            This is a convenience function that calls the timeit()
            repeatedly, returning a list of results.  The first argument
            specifies how many times to call timeit(), defaulting to 5;
            the second argument specifies the timer argument, defaulting
            to one million.

            Note: it's tempting to calculate mean and standard deviation
            from the result vector and report these.  However, this is not
            very useful.  In a typical case, the lowest value gives a
            lower bound for how fast your machine can run the given code
            snippet; higher values in the result vector are typically not
            caused by variability in Python's speed, but by other
            processes interfering with your timing accuracy.  So the min()
            of the result is probably the only number you should be
            interested in.  After that, you should look at the entire
            vector and apply common sense rather than statistics.
            """
        r = []
        for _ in range(repeat):
            t = self.timeit(number)
            r.append(t)
        return r

    def autorange(self, callback=None) -> t.Tuple[int, float]:
        """ Return the number of loops and time taken so that total time >= 0.2.

            Calls the timeit method with increasing numbers from the sequence
            1, 2, 5, 10, 20, 50, ... until the time taken is at least 0.2
            second.  Returns (number, time_taken).

            If *callback* is given and is not None, it will be called after
            each trial with two arguments: ``callback(number, time_taken)``.
            """
        i: int = 1
        j: int
        number: int
        time_taken: float
        while True:
            for j in 1, 2, 5:
                number = i * j
                time_taken = self.timeit(number)
                if callback:
                    callback(number, time_taken)
                if time_taken >= 0.2:
                    return (number, time_taken)
            i *= 10


def timeit(func: t.Callable = _pass,
           setup: t.Callable = _pass,
           timer: time.perf_counter = default_timer,
           number: int = default_number,
           globals: t.Dict[str, t.Any] = None) -> float:
    """Convenience function to create Timer object and call timeit method."""
    return Timer(func, setup, timer, globals).timeit(number)


def repeat(func: t.Callable = _pass,
           setup: t.Callable = _pass,
           timer: time.perf_counter = default_timer,
           number: int = default_number,
           repeat: int = default_repeat,
           globals: t.Dict[str, t.Any] = None) -> t.List[float]:
    """Convenience function to create Timer object and call repeat method."""
    return Timer(func, setup, timer, globals).repeat(repeat, number)


if __name__ == "__main__":
    from timebandit._cli import main
    sys.exit(main())
